### DISCUSSION TITLE
How to 'Quit' an app (default Processing window)

### DISCUSSION THREAD

**Discussion by vsquared** [py5coding/py5generator#461] [Q&A] [answered]
Currently the only method to quit an app on a Mac is to use the close window button at the top left corner of the window.  Normally there would be a cmd-Q menu item under the app menu on the menubar.  There is a default menubar item called 'Python' but there are no menu items below it where the 'Quit' menu item would normally go.  In addition, I usually put a 'Quit' button on the main window canvas.  If I use a custom button and call _exit()_ when pressed (which works in Processing), it gives the following error:

```
org.jpype.PyExceptionProxy
	at org.jpype.JPypeContext.createException(Unknown Source)
	at org.jpype.proxy.JPypeProxy.hostInvoke(Native Method)
	at org.jpype.proxy.JPypeProxy.invoke(Unknown Source)
	at jdk.proxy2/jdk.proxy2.$Proxy6.run_method(Unknown Source)
	at py5.core.Sketch.handleInputEventNoParams(Sketch.java:258)
	at py5.core.Sketch.mousePressed(Sketch.java:271)
	at py5.core.Sketch.mousePressed(Sketch.java:265)
	at processing.core.PApplet.handleMouseEvent(PApplet.java:2371)
	at processing.core.PApplet.dequeueEvents(PApplet.java:2261)
	at processing.core.PApplet.handleDraw(PApplet.java:2104)
	at processing.awt.PSurfaceAWT$9.callDraw(PSurfaceAWT.java:1386)
	at processing.core.PSurfaceNone$AnimationThread.run(PSurfaceNone.java:356)

Process ended with exit code 0.
``` 
If I use a JButton with a handler, there is a similar more lengthy error.  Is there a better way of doing this or should I call a different method?  The menubar is shown below.  If I could programmatically add a 'Quit' menu item to the menubar that would also be nice if we have access to it.
<img width="107" alt="mbar" src="https://github.com/py5coding/py5generator/assets/2294268/07e4957a-2682-4f48-8801-c2eea5a74ea1">

**Comment by hx2A** [comment, marked as answer]
Can you try `exit_sketch()`?

https://py5coding.org/reference/sketch_exit_sketch.html

If your button called that method it might do what you need it to do.

py5 has no `exit()` method because it would conflict with the Python builtin function. Instead it is renamed to `exit_sketch()`.

I just checked our [Tips for Processing Java Users](https://py5coding.org/content/coming_from_processing_java.html) page and it does not mention the exit/exit_sketch rename. I'll open an issue so I remember to add that later.

**Reply by vsquared** [reply]
exit_sketch() works without error when used with a custom button in default window.  Thanks for that and also the helpful links.

**Reply by hx2A** [reply]
Great, glad we could help!

**Reply by villares** [reply]
> I just checked our [Tips for Processing Java Users](https://py5coding.org/content/coming_from_processing_java.html) page and it does not mention the exit/exit_sketch rename. I'll open an issue so I remember to add that later.

Just a note to register that the Tips page has been updated to include `exit_sketch()`, yay!
